docs/memorymanagement.md

```diff
@@ -0,0 +1,114 @@
+implemented on the outer object or is overridden by the outer object or is for the `IUnknown` interface
+would be handled by the outer object itself. The last part means that the lifetime and the COM reference
+counting of this aggregated object is maintained by the outer object and more specifically its `IUnknown`
+implementation on the CCW from ComWrappers. This is where the standard COM reference tracking
+convention described earlier starts to differ. As we know for CCWs, there are 2 things which
+
+keep it alive: any references from C# to the managed object or any native references which had done
+an `AddRef` incrementing the COM ref count. But we also know that for projected aggregated types
+to make calls on interfaces provided by the inner object, they need to QueryInterface (QI) for them
+from the inner object which would result in the COM reference count on the outer (CCW) increasing.
+This means any QIs from C# on such objects will end up increasing the COM reference count on the CCW
+and thereby keeping it alive and leaking it as any C# reference to such objects are
+supposed to be tracked as managed references by the garbage collector and not as native references.
+To address this, for any QI calls done as part of the aggregated object's C# projection implementation,
+`Release` should be called right after the reference is obtained even if you plan to hold onto
+the obtained interface to avoid repeatedly retrieving it. This prevents C# QIs by the composed object
+from increasing the CCW reference count meant for tracking native references while allowing the
+garbage collector to manage the lifetime of managed objects from managed references via its own
+tracking. For any QI calls for which the result is handed out to the native side, `Release` should
+not be called right after as it is a native reference which needs to be tracked by the CCW.
+
+One notable caveat to this is tear off interfaces on aggregated objects. With tear off interfaces,
+the interfaces typically perform their own COM reference counting separate from the object itself
+allowing for the interfaces to manage their own lifetime. But this doesn't work well with aggregated
+objects if one of these interfaces need to be QIed for by the composed object as part of the
+projection implementation. This is because a `Release` would happen right after which would trigger
+the cleanup of the interface as its lifetime isn't tied to the outer. Given that tear off interfaces
+are rare and not typically used by C# consumers, C#/WinRT today doesn't address this
+other than facilitating QI calls for them from the native side where `Release` isn't called right after.
+The recommendation for tear off interfaces which do want to support such uses on aggregated objects
+is that they can continue to be constructed on demand upon the first QI for it, but the interface
+should not be cleaned up until the object is cleaned up even if there are no longer any reference

```

Comment:
references

Reply:
can we be more crisp about the recommendation - that a tear-off implementor should cache all instances to prevent premature destruction?